File: lib/src/domain/standings_calculator.dart
import 'bracket_rules.dart';
import 'models.dart';

class StandingsCalculator {
  const StandingsCalculator();

  static const _maxFormLength = 5;

  List<GroupStanding> calculate({
    required Iterable<Fixture> fixtures,
    required Map<String, List<String>> overrideOrdersByGroup,
    DateTime? updatedAt,
    String? updatedBy,
  }) {
    return [
      for (final groupId in BracketRules.groupIds)
        _calculateGroup(
          groupId: groupId,
          fixtures: fixtures,
          overrideOrderCountryIds: overrideOrdersByGroup[groupId] ?? const [],
          updatedAt: updatedAt,
          updatedBy: updatedBy,
        ),
    ];
  }

  GroupStanding _calculateGroup({
    required String groupId,
    required Iterable<Fixture> fixtures,
    required List<String> overrideOrderCountryIds,
    DateTime? updatedAt,
    String? updatedBy,
  }) {
    final countryIds = BracketRules.groupCountryIds[groupId] ?? const [];
    final stats = {
      for (final countryId in countryIds)
        countryId: _MutableStanding(countryId),
    };
    final groupFixtures =
        fixtures
            .where(
              (fixture) =>
                  fixture.stage == TournamentStage.group &&
                  fixture.status == FixtureStatus.finished &&
                  fixture.roundLabel.trim().toUpperCase() == 'GROUP $groupId' &&
                  fixture.homeCountryId != null &&
                  fixture.awayCountryId != null &&
                  fixture.homeScore != null &&
                  fixture.awayScore != null,
            )
            .toList()
          ..sort((a, b) => a.kickoff.compareTo(b.kickoff));

    for (final fixture in groupFixtures) {
      final home = stats[fixture.homeCountryId];
      final away = stats[fixture.awayCountryId];
      if (home == null || away == null) {
        continue;
      }
      home.apply(
        goalsFor: fixture.homeScore!,
        goalsAgainst: fixture.awayScore!,
      );
      away.apply(
        goalsFor: fixture.awayScore!,
        goalsAgainst: fixture.homeScore!,
      );
    }

    final rows =
        stats.values.map((standing) => standing.toRow()).toList()
          ..sort(_defaultCompare);
    final orderedRows = _applyOverride(rows, overrideOrderCountryIds);
    return GroupStanding(
      groupId: groupId,
      rows: [
        for (var index = 0; index < orderedRows.length; index++)
          orderedRows[index].copyWith(rank: index + 1),
      ],
      overrideOrderCountryIds: overrideOrderCountryIds,
      updatedAt: updatedAt,
      updatedBy: updatedBy,
    );
  }

  int _defaultCompare(StandingRow a, StandingRow b) {
    if (b.points != a.points) return b.points - a.points;
    if (b.goalDifference != a.goalDifference) {
      return b.goalDifference - a.goalDifference;
    }
    if (b.goalsFor != a.goalsFor) return b.goalsFor - a.goalsFor;
    return a.countryId.compareTo(b.countryId);
  }

  List<StandingRow> _applyOverride(
    List<StandingRow> rows,
    List<String> overrideOrderCountryIds,
  ) {
    if (overrideOrderCountryIds.isEmpty) {
      return rows;
    }
    final byCountryId = {for (final row in rows) row.countryId: row};
    return [
      for (final countryId in overrideOrderCountryIds)
        if (byCountryId[countryId] != null) byCountryId.remove(countryId)!,
      ...byCountryId.values,
    ];
  }
}

class _MutableStanding {
  _MutableStanding(this.countryId);

  final String countryId;
  int played = 0;
  int won = 0;
  int drawn = 0;
  int lost = 0;
  int goalsFor = 0;
  int goalsAgainst = 0;
  final List<String> _recentResults = [];

  int get goalDifference => goalsFor - goalsAgainst;
  int get points => won * 3 + drawn;

  void apply({required int goalsFor, required int goalsAgainst}) {
    played++;
    this.goalsFor += goalsFor;
    this.goalsAgainst += goalsAgainst;
    if (goalsFor > goalsAgainst) {
      won++;
      _recentResults.add('W');
    } else if (goalsFor < goalsAgainst) {
      lost++;
      _recentResults.add('L');
    } else {
      drawn++;
      _recentResults.add('D');
    }
    while (_recentResults.length > StandingsCalculator._maxFormLength) {
      _recentResults.removeAt(0);
    }
  }

  String get form {
    if (_recentResults.isEmpty) {
      return '';
    }
    return _recentResults.join(',');
  }

  StandingRow toRow() {
    return StandingRow(
      countryId: countryId,
      rank: 0,
      played: played,
      won: won,
      drawn: drawn,
      lost: lost,
      goalsFor: goalsFor,
      goalsAgainst: goalsAgainst,
      goalDifference: goalDifference,
      points: points,
      form: form,
    );
  }
}

File: functions/src/standings-calculator.ts
import {groupCountryIds, groupIds} from "./bracket-rules";

export type FixtureRecord = {
  id: string;
  stage?: string;
  status?: string;
  roundLabel?: string;
  kickoff?: string;
  homeCountryId?: string | null;
  awayCountryId?: string | null;
  homeScore?: number | null;
  awayScore?: number | null;
};

export type StandingRow = {
  countryId: string;
  rank: number;
  played: number;
  won: number;
  drawn: number;
  lost: number;
  goalsFor: number;
  goalsAgainst: number;
  goalDifference: number;
  points: number;
  form: string;
};

export type GroupStanding = {
  groupId: string;
  rows: StandingRow[];
  overrideOrderCountryIds: string[];
  updatedAt?: string;
  updatedBy?: string;
};

type MutableStanding = {
  countryId: string;
  played: number;
  won: number;
  drawn: number;
  lost: number;
  goalsFor: number;
  goalsAgainst: number;
  recentResults: string[];
};

const MAX_FORM_LENGTH = 5;

export function calculateStandings(
  fixtures: FixtureRecord[],
  existingStandings: Array<{
    groupId: string;
    overrideOrderCountryIds?: string[];
  }> = [],
  updatedAt?: string,
  updatedBy?: string,
): GroupStanding[] {
  const overrides = new Map(
    existingStandings.map((standing) => [
      standing.groupId,
      standing.overrideOrderCountryIds ?? [],
    ]),
  );

  return groupIds.map((groupId) =>
    calculateGroupStanding(
      groupId,
      fixtures,
      overrides.get(groupId) ?? [],
      updatedAt,
      updatedBy,
    ),
  );
}

function calculateGroupStanding(
  groupId: string,
  fixtures: FixtureRecord[],
  overrideOrderCountryIds: string[],
  updatedAt?: string,
  updatedBy?: string,
): GroupStanding {
  const countryIds = groupCountryIds[groupId] ?? [];
  const stats = new Map<string, MutableStanding>(
    countryIds.map((countryId) => [
      countryId,
      {
        countryId,
        played: 0,
        won: 0,
        drawn: 0,
        lost: 0,
        goalsFor: 0,
        goalsAgainst: 0,
        recentResults: [],
      },
    ]),
  );

  const groupFixtures = fixtures
    .filter(
      (fixture) =>
        fixture.stage === "group" &&
        fixture.status === "finished" &&
        fixture.roundLabel?.trim().toUpperCase() === `GROUP ${groupId}` &&
        fixture.homeCountryId &&
        fixture.awayCountryId &&
        fixture.homeScore != null &&
        fixture.awayScore != null,
    )
    .sort((a, b) => {
      const aTime = a.kickoff ? Date.parse(a.kickoff) : 0;
      const bTime = b.kickoff ? Date.parse(b.kickoff) : 0;
      return aTime - bTime;
    });

  for (const fixture of groupFixtures) {
    const home = stats.get(fixture.homeCountryId!);
    const away = stats.get(fixture.awayCountryId!);
    if (!home || !away) {
      continue;
    }

    applyResult(home, fixture.homeScore!, fixture.awayScore!);
    applyResult(away, fixture.awayScore!, fixture.homeScore!);
  }

  const rows = [...stats.values()]
    .map((standing) => toRow(standing))
    .sort(defaultCompare);
  const orderedRows = applyOverride(rows, overrideOrderCountryIds);

  return {
    groupId,
    rows: orderedRows.map((row, index) => ({...row, rank: index + 1})),
    overrideOrderCountryIds,
    updatedAt,
    updatedBy,
  };
}

function applyResult(
  standing: MutableStanding,
  goalsFor: number,
  goalsAgainst: number,
): void {
  standing.played += 1;
  standing.goalsFor += goalsFor;
  standing.goalsAgainst += goalsAgainst;
  if (goalsFor > goalsAgainst) {
    standing.won += 1;
    standing.recentResults.push("W");
  } else if (goalsFor < goalsAgainst) {
    standing.lost += 1;
    standing.recentResults.push("L");
  } else {
    standing.drawn += 1;
    standing.recentResults.push("D");
  }
  while (standing.recentResults.length > MAX_FORM_LENGTH) {
    standing.recentResults.shift();
  }
}

function toRow(standing: MutableStanding): StandingRow {
  return {
    countryId: standing.countryId,
    rank: 0,
    played: standing.played,
    won: standing.won,
    drawn: standing.drawn,
    lost: standing.lost,
    goalsFor: standing.goalsFor,
    goalsAgainst: standing.goalsAgainst,
    goalDifference: standing.goalsFor - standing.goalsAgainst,
    points: standing.won * 3 + standing.drawn,
    form: standing.recentResults.join(","),
  };
}

function defaultCompare(a: StandingRow, b: StandingRow): number {
  if (b.points !== a.points) return b.points - a.points;
  if (b.goalDifference !== a.goalDifference) {
    return b.goalDifference - a.goalDifference;
  }
  if (b.goalsFor !== a.goalsFor) return b.goalsFor - a.goalsFor;
  return a.countryId.localeCompare(b.countryId);
}

function applyOverride(
  rows: StandingRow[],
  overrideOrderCountryIds: string[],
): StandingRow[] {
  if (overrideOrderCountryIds.length === 0) {
    return rows;
  }
  const byCountryId = new Map(rows.map((row) => [row.countryId, row]));
  const ordered: StandingRow[] = [];
  for (const countryId of overrideOrderCountryIds) {
    const row = byCountryId.get(countryId);
    if (row) {
      ordered.push(row);
      byCountryId.delete(countryId);
    }
  }
  ordered.push(...byCountryId.values());
  return ordered;
}
